README.txt
                    Bitnami Drupal Stack 7.41-1
                  ============================


1. OVERVIEW

The Bitnami Project was created to help spread the adoption of freely
available, high quality, open source web applications. Bitnami aims to make
it easier than ever to discover, download and install open source software
such as document and content management systems, wikis and blogging
software.

You can learn more about Bitnami at https://bitnami.com

Drupal is a content management platform that allows an individual or community 
of users to easily publish, manage, and organize a wide variety of content on a 
website. In general, Drupal is used for community web portals, discussion sites, 
corporate web sites, intranet applications, personal web sites or blogs, 
aficionado sites, e-commerce applications, resource directories, and social 
networking sites. Drupal is easy to extend by plugging in one or more of the 
dozens of freely available modules.

You can learn more about Drupal at http://drupal.org

The Bitnami Drupal Stack is an installer that greatly simplifies the
installation of Drupal and runtime dependencies. It includes ready-to-run
versions of Apache, MySQL and PHP. Drupal Stack is distributed for free
under the Apache 2.0 license. Please see the appendix for the specific
licenses of all open source components included.

You can learn more about Bitnami Stacks at https://bitnami.com/stacks/

2. FEATURES

- Easy to Install

Bitnami Stacks are built with one goal in mind: to make it as easy as
possible to install open source software. Our installers completely automate
the process of installing and configuring all of the software included in
each Stack, so you can have everything up and running in just a few clicks.

- Independent

Bitnami Stacks are completely self-contained, and therefore do not interfere
with any software already installed on your system. For example, you can
upgrade your system's MySQL or Apache without fear of 'breaking' your
Bitnami Stack.

- Integrated

By the time you click the 'finish' button on the installer, the whole stack
will be integrated, configured and ready to go. 

- Relocatable

Bitnami Stacks can be installed in any directory. This allows you to have
multiple instances of the same stack, without them interfering with each other. 

3. COMPONENTS

Bitnami Drupal Stack ships with the following:

  - Drupal 7.41
  - Apache 2.4.17
  - MySQL 5.6.26
  - PHP 5.5.30
  - phpMyAdmin 4.4.13.1
  - drush 6.5.0

4. REQUIREMENTS

To install Bitnami Drupal Stack you will need:

    - Intel x86 or compatible processor
    - Minimum of 256 MB RAM 
    - Minimum of 150 MB hard drive space
    - TCP/IP protocol support
    - Compatible operantig systems:
      - An x86 Linux operating system.
      - A 32-bit Windows operating system such as Windows Vista, Windows 7, Windows 8, Windows 10, Windows Server 2008 or Windows Server 2012.
			- An OS X operating system x86.

4. INSTALLATION

The Bitnami Drupal Stack is distributed as a binary executable installer.
It can be downloaded from:

https://bitnami.com/stacks/

The downloaded file will be named something similar to:

bitnami-drupal-7.41-1-linux-installer.run on Linux or
bitnami-drupal-7.41-1-linux-x64-installer.run on Linux or
bitnami-drupal-7.41-1-windows-installer.exe on Windows or
bitnami-drupal-7.41-1-osx-x86-installer.dmg on OS X x86.

On Linux, you will need to give it executable permissions:

chmod 755 bitnami-drupal-7.41-1-linux.run


To begin the installation process, double-click on that file, and you will
be greeted by the 'Welcome' screen. Pressing 'Next' will take you to the
Component Selection screen, where you can select the phpMyAdmin component. This
tool intended to handle the administration of MySQL over the web.

The next screen is the Installation Folder, where you can select where Bitnami 
stack will be installed. If the destination directory does not exist, it will 
be created as part of the installation. 

The next screen will prompt you for data necessary to create the initial
admin user: 

Username and password: You will use this information to log-in into the
administrative interface. The password you provide here will also be used to
protect other parts of the installation. Please see the section named
"Usernames and Passwords" later in this document.

Real name: The name that will be displayed in the application.

Email address: Your email address.

The next screen will vary, depending on whether the ports needed by the
bundled Apache and MySQL are already taken. The default listening port for
Apache is 8080 on Linux and 80 on Wndows and for MySQL is 3006. If those ports 
are already in use by other applications, you will be prompted for alternate 
ports to use.

Once the information has been entered, the installation will proceed to copy
the files to the target installation directory and configure the different
components. One this process has  been completed, you will see the
'Installation Finished' page. You can choose to launch Bitnami Drupal
Stack at this point. If you do so, your default web browser will open and
display the Welcome page for the Bitnami Drupal Stack. 

If you received an error message during installation, please refer to
the Troubleshooting section.

The rest of this guide assumes that you installed Bitnami Drupal
Stack in /home/user/drupal-7.41-1 on Linux or C:\Program Files\Bitnami Drupal Stack
on Windows or /Applications/drupal-7.41-1 on OS X and that you use port 
8080 for Apache on Linux and port 80 on Windows and 3306 for MySQL. 

5. STARTING AND STOPPING BITNAMI DRUPAL STACK

To start/stop/restart application on Linux you can use the included ctlscript.sh
utility, as shown below:

       ./ctlscript.sh (start|stop|restart)
       ./ctlscript.sh (start|stop|restart) mysql
       ./ctlscript.sh (start|stop|restart) apache

  start      - start the service(s)
  stop       - stop  the service(s)
  restart    - restart or start the service(s)

The Bitnami Control Panel is a simple graphical interface included in the OS X stacks that 
can start and stop the Bitnami servers. It is are located in the same installation directory. 
To start the utility, double click the file named ‘control’ from your file browser.

On Windows, you can start and stop Bitnami Drupal Stack on Windows using the shortcuts created
in the Start Menu, under Programs -> Bitnami -> Bitnami Service 

That will start Apache service. Once started, you can open your
browser and access the following URL on Linux or OS X:

http://127.0.0.1:8080/drupal/

And you can access the following URL on Windows:

http://127.0.0.1/drupal/

If you selected an alternate port during installation, for example 18080, the
URL will look like:

http://127.0.0.1:18080/drupal/

If you are accessing the machine remotely, you will need to replace
127.0.0.1 with the appropriate IP address or hostname.


6. DIRECTORY STRUCTURE

The installation process will create several subfolders under the main
installation directory:

	apache2/: Apache Web server.
	php/: PHP Scripting Language.
	mysql/: MySQL Database.
	apps/
	  drupal/: Drupal application folder
	    conf/: Drupal Apache configuration files
	    htdocs/: Drupal PHP application files
	  phpMyAdmin/: phpMyAdmin application folder (optional)

7. DEFAULT USERNAMES AND PASSWORDS

The Drupal administrative user and password are the ones you set at
installation time. 

MySQL admin user is called 'root', and its password is the same as the
Drupal administrative user password.

The default MySQL non-root account used to access the database is named
bitnami, and its password is also bitnami. 

The phpMyAdmin administrative user is 'root'.

8. TROUBLESHOOTING

This version of the Bitnami Drupal stack is a preview version, and as
such, may contain a number of bugs and be a little bit rough around the
edges. We are working on the next release, which will contain several
improvements along with expanded documentation. In addition to the resources
provided below, we encourage you to post your questions and suggestions at:

https://community.bitnami.com/

We also encourage you to sign up for our newsletter, which we'll use to
announce new releases and new stacks. To do so, just register at:
https://bitnami.com/newsletter.  

8.1 Installer

# Installer Payload Error 

If you get the following error while trying to run the installer from the
command line:

"Installer payload initialization failed. This is likely due to an
incomplete or corrupt downloaded file" 

The installer binary is not complete, likely because the file was
not downloaded correctly. You will need to download the file and
repeat the installation process. 

# Installer execution error on Linux

If you get the following error while trying to run the installer:

"Cannot open bitnami-drupal-7.41-1-linux.run: No application suitable for
automatic installation is available for handling this kind of file."

In some operatings systems you can change permissions with right click
properties -> permissions -> execution enable.

Or from the command line:

$ chmod 755 bitnami-drupal-7.41-1-linux.run

8.2 Apache

If you find any problem starting Apache, the first place you should check is
the Apache error log file:

/home/user/drupal/apache2/logs/error.log on Linux or
C:\Program Files\Bitnami Drupal Stack\apache2\logs\error.log on Windows or
/Applications/drupal-7.41-1/apache2/logs/error.log.

Most errors are related to not being able to listen to the default port.
Make sure there are no other server programs listening at the same port
before trying to start Apache. Some programs, such as Skype, may also use
port 80. For issues not covered in this Quick Start guide, please refer to
the Apache documentation, which is located at http://httpd.apache.org/docs/  

8.3 MySQL

The following are some common problems: 

# Access denied when trying to connect to MySQL. 

If you get an Access Denied message while trying to connect to
MySQL, make sure you are using the correct username and password. 

# "Can't connect to server" message. 

Make sure that the MySQL daemon is up and running. Remember also that if
during installation you selected a different listening port for MySQL, you
may need to pass that as an extra command line option.

For help in troubleshooting MySQL errors, you may want to reference the
"Problems and Common Errors" section of the MySQL manual, which you will
find at http://dev.mysql.com/doc/


8.4 Drupal

For any problem related to Drupal, please visit 
http://drupal.org/support


9. LICENSES

Apache Web Server is distributed under the Apache License v2.0, which
is located at http://www.apache.org/licenses/LICENSE-2.0

MySQL is distributed under the GNU General Public License v2, which is
located at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

PHP and related libraries are distributed under the PHP License v3.01,
which is located at http://www.php.net/license/3_01.txt

Drupal is distributed under the GNU General Public License v2,
which is located at
http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

curl is distributed under the Curl License, which is located at
http://curl.haxx.se/docs/copyright.html

expat is distributed under the MIT License, which is located at
http://www.jclark.com/xml/copying.txt

gd is distributed under the gd License, which is located at
http://www.boutell.com/gd/manual2.0.33.html

IMAP is distributed under the University of Washington Free-Fork
License, located at http://www.washington.edu/imap/legal.html

jpegsrc is distributed under The Independent JPEG Group's JPEG
software license, which is located at
http://dev.w3.org/cvsweb/Amaya/libjpeg/README?rev=1.2

libiconv is distributed under the Lesser General Public License
(LGPL), located at http://www.gnu.org/copyleft/lesser.html

Freetype is released under The Freetype Project License, that is located at
http://freetype.sourceforge.net/FTL.TXT


changelog.txt
============ CHANGELOG =============

Version 7.41-1      2015-11-16
* Updated libpng to 1.5.24 (security issue)

Version 8.0.0.rc3-0      2015-11-05
* Updated Drupal to 8.0.0-rc3
* Updated MySQL to 5.6.27

Version 8.0.0.rc2-0      2015-10-22
* Updated Drupal to 8.0.0-rc2

Version 7.41-0      2015-10-22
* Updated Drupal to 7.41

Version 7.40-0      2015-10-15
* Updated Drupal to 7.40
* Updated Apache to 2.4.17

Version 8.0.0.rc1-1      2015-10-13
* Added Windows and module support for  Drupal 8.0.0.rc1 

Version 8.0.0.rc1-0      2015-09-04
* Updated Drupal to 8.0.0-rc1
* Updated PHP to 5.5.29

Version 8.0.0.beta15-0      2015-09-04
* Updated Drupal to 8.0.0.beta15

Version 7.39-0      2015-08-20
* Updated Drupal to 7.39 (security issues)

Version 6.37-0      2015-08-20
* Updated Drupal to 6.37 (security issues)
* Updated phpMyAdmin to 4.4.13.1
* Updated Apache to 2.4.16
* Updated PHP to 5.5.28

Version 8.0.0.beta14-0      2015-08-04
* Updated Drupal to 8.0.0.beta14

Version 8.0.0.beta13-0      2015-07-31
* Updated Drupal to 8.0.0.beta13
* Fix for browser launch in Windows 10 Edge browser

Version 7.38-2      2015-07-31
* Fix for browser launch in Windows 10 Edge browser

Version 8.0.0.beta12-1      2015-07-10
* Updated OpenSSL to 1.0.1p

Version 7.38-1      2015-07-10
* Updated OpenSSL to 1.0.1p

Version 6.36-1      2015-07-10
* Updated OpenSSL to 1.0.1p

Version 8.0.0.beta12-0      2015-07-03
* Updated Drupal to 8.0.0.beta12

Version 8.0.0.beta11-0      2015-06-20
* Updated Drupal to 8.0.0.beta11

Version 6.36-0      2015-06-18
* Updated Drupal to 6.36 (security issues)

Version 7.38-0      2015-06-18
* Updated Drupal to 7.38 (security issues)
* Updated PHP to 5.5.26
* Updated MySQL to 5.6.25
* Updated PHP to 5.5.25
* Updated phpMyAdmin to 4.4.7
* Updated Drupal to 7.37

Version 8.0.0.beta10-1      2015-05-07
* Using Drupal official repository instead of GitHub repository
* Dropping RC4 from Apache SSLCipherSuite directive

Version 8.0.0.beta10-0      2015-05-01
* Updated Drupal to 8.0.0.beta10
* Updated phpMyAdmin to 4.4.3
* Updated PHP to 5.4.40

Version 8.0.0.beta9-0      2015-04-08
* Updated Drupal to 8.0.0.beta9
* Updated OpenSSL to 1.0.1m
* Updated PHP to 5.4.39
* Fixed zlib1.dll issue with Drush on Windows

Version 7.36-0      2015-04-02
* Updated Drupal to 7.36
* Updated OpenSSL to 1.0.1m
* Updated PHP to 5.4.39
* Fixed zlib1.dll issue with Drush on Windows

Version 7.35-0      2015-03-19
* Updated Drupal to 7.35

Version 6.35-0      2015-03-19
* Updated Drupal to 6.35
* Updated OpenSSL to 1.0.1l
* Fixed binaries for Git

Version 8.0.0.beta7-0      2015-02-27
* Updated PHP to 5.4.38
* Updated Drupal8 to 8.0.0.beta7
* Updated Apache to 2.4.12
* Updated MySQL to 5.5.42 for Linux and Windows
* Updated PageSpeed to 1.9.32.3

Version 8.0.0.beta6-0      2015-01-29
* Updated Drupal8 to 8.0.0.beta6

Version 8.0.0.beta4-0      2015-01-09    
* Updated phpMyAdmin to 4.3.7
* Updated OpenSSL to 1.0.1k (Linux and OS X)
* Updated Drupal8 to 8.0.0.beta4

Version 7.34-1      2014-12-26
* Updated phpMyAdmin to 4.3.3
* Updated PHP to 5.4.36
* Updated Git to 1.9.5
* Updated PHP to 5.4.35

Version 7.34-0      2014-11-20
* Updated Drupal to 7.34

Version 6.34-0      2014-11-20
* Updated Drupal to 6.34

Version 8.0.0.beta3-0      2014-11-15
* Updated Drupal 8 to 8.0.0.beta3

Version 7.33-0      2014-11-12
* Updated Drupal to 7.33

Version 6.33-2      2014-10-22
* Fix MySQL startup issue for OS X Yosemite
* Updated PHP to 5.4.34
* Updated OpenSSL to 1.0.1j
* Added OCI8 module. It requires InstantClient 11.2.
* Signed with codesign v2
* Updated phpMyAdmin to 4.2.9.1
* Updated MySQL to 5.5.40

Version 8.0.0.alpha14-0      2014-10-22
* (Version only available for OS X)
* Fix MySQL startup issue for OS X Yosemite
* Updated PHP to 5.4.34
* Updated OpenSSL to 1.0.1j

Version 7.32-0      2014-10-22
* (Version only available for OS X)
* Fix MySQL startup issue for OS X Yosemite
* Updated PHP to 5.4.34
* Updated OpenSSL to 1.0.1j

Version 7.32-0      2014-10-16
* Updated Drupal to 7.32
* Added OCI8 module. It requires InstantClient 11.2.
* Signed with codesign v2

Version 7.31-2      2014-10-08
* Updated phpMyAdmin to 4.2.9.1
* Updated MySQL to 5.5.40
* Signed with codesign v2
* Updated PHP to 5.4.33

Version 8.0.0.alpha14-0      2014-08-27
* Updated Drupal to 8.0.0-alpha14
* Moved application to root URL in Cloud Images and Virtual Machines

Version 6.33-1      2014-08-27
* Moved application to root URL in Cloud Images and Virtual Machines

Version 7.31-1      2014-08-27
* Moved application to root URL in Cloud Images and Virtual Machines
* Updated MySQL to 5.5.39
* Updated PHP to 5.4.32
* Updated OpenSSL to 1.0.1i
* Updated phpMyAdmin to 4.2.7.1

Version 6.33-0      2014-08-12
* Updated Drupal to 6.33

Version 7.31-0      2014-08-11
* Updated Drupal to 7.31

Version 7.30-0      2014-07-25
* Updated Apache to 2.4.10
* Updated PHP to 5.4.31

Version 7.29-0      2014-07-17
* Updated MySQL to 5.5.38
* Updated XDebug to 2.2.5
* Updated XCache to 3.1.0
* Updated mongo-php-driver to 1.5.4
* Updated php-couchbase to 1.1.5
* Updated Libxslt to 1.1.28
* Updated Libxml2 to 2.9.1

Version 6.32-0      2014-07-17
* Updated Drupal to 6.32
* Updated MySQL to 5.5.38
* Updated XDebug to 2.2.5
* Updated XCache to 3.1.0
* Updated mongo-php-driver to 1.5.4
* Updated php-couchbase to 1.1.5
* Updated Libxslt to 1.1.28
* Updated Libxml2 to 2.9.1

Version 8.0.alpha13-0      2014-07-05
* Updated PHP to 5.4.30
* Updated phpMyAdmin to 4.2.5
* Updated OpenSSL to 1.0.1h
* Updated PHP to 5.4.29
* Fixed module installation issue

Version 8.0.alpha12-0      2014-05-29
* Updated Drupal to 8.0.alpha12
* Updated phpMyAdmin to 4.2.2

Version 8.x.20140513-0      2014-05-14
* Updated Drush for Drupal 8 to master

Version 7.28-0      2014-05-08
* Updated Drupal to 7.28
* Added libv8 and v8js PHP extension for Linux and OS X

Version 8.0.alpha11-0      2014-04-24
* Updated Drupal to 8.0.alpha11

Version 7.27-0      2014-04-21

php/lib/php/pearcmd.php
<?php
/**
 * PEAR, the PHP Extension and Application Repository
 *
 * Command line interface
 *
 * PHP versions 4 and 5
 *
 * @category   pear
 * @package    PEAR
 * @author     Stig Bakken <ssb@php.net>
 * @author     Tomas V.V.Cox <cox@idecnet.com>
 * @copyright  1997-2009 The Authors
 * @license    http://opensource.org/licenses/bsd-license.php New BSD License
 * @version    CVS: $Id: pearcmd.php 313023 2011-07-06 19:17:11Z dufuz $
 * @link       http://pear.php.net/package/PEAR
 */

ob_end_clean();
if (!defined('PEAR_RUNTYPE')) {
    // this is defined in peclcmd.php as 'pecl'
    define('PEAR_RUNTYPE', 'pear');
}
define('PEAR_IGNORE_BACKTRACE', 1);
/**
 * @nodep Gtk
 */
if ('/Applications/drupal-7.41-1/php/lib/php' != '@'.'include_path'.'@') {
    ini_set('include_path', '/Applications/drupal-7.41-1/php/lib/php');
    $raw = false;
} else {
    // this is a raw, uninstalled pear, either a cvs checkout, or php distro
    $raw = true;
}
@ini_set('allow_url_fopen', true);
if (!ini_get('safe_mode')) {
    @set_time_limit(0);
}
ob_implicit_flush(true);
@ini_set('track_errors', true);
@ini_set('html_errors', false);
@ini_set('magic_quotes_runtime', false);
$_PEAR_PHPDIR = '#$%^&*';
set_error_handler('error_handler');

$pear_package_version = "1.9.4";

require_once 'PEAR.php';
require_once 'PEAR/Frontend.php';
require_once 'PEAR/Config.php';
require_once 'PEAR/Command.php';
require_once 'Console/Getopt.php';


PEAR_Command::setFrontendType('CLI');
$all_commands = PEAR_Command::getCommands();

// remove this next part when we stop supporting that crap-ass PHP 4.2
if (!isset($_SERVER['argv']) && !isset($argv) && !isset($HTTP_SERVER_VARS['argv'])) {
    echo 'ERROR: either use the CLI php executable, or set register_argc_argv=On in php.ini';
    exit(1);
}

$argv = Console_Getopt::readPHPArgv();
// fix CGI sapi oddity - the -- in pear.bat/pear is not removed
if (php_sapi_name() != 'cli' && isset($argv[1]) && $argv[1] == '--') {
    unset($argv[1]);
    $argv = array_values($argv);
}
$progname = PEAR_RUNTYPE;
array_shift($argv);
$options = Console_Getopt::getopt2($argv, "c:C:d:D:Gh?sSqu:vV");
if (PEAR::isError($options)) {
    usage($options);
}

$opts = $options[0];

$fetype = 'CLI';
if ($progname == 'gpear' || $progname == 'pear-gtk') {
    $fetype = 'Gtk';
} else {
    foreach ($opts as $opt) {
        if ($opt[0] == 'G') {
            $fetype = 'Gtk';
        }
    }
}
//Check if Gtk and PHP >= 5.1.0
if ($fetype == 'Gtk' && version_compare(phpversion(), '5.1.0', '>=')) {
    $fetype = 'Gtk2';
}

$pear_user_config = '';
$pear_system_config = '';
$store_user_config = false;
$store_system_config = false;
$verbose = 1;

foreach ($opts as $opt) {
    switch ($opt[0]) {
        case 'c':
            $pear_user_config = $opt[1];
            break;
        case 'C':
            $pear_system_config = $opt[1];
            break;
    }
}

PEAR_Command::setFrontendType($fetype);
$ui = &PEAR_Command::getFrontendObject();
$config = &PEAR_Config::singleton($pear_user_config, $pear_system_config);

if (PEAR::isError($config)) {
    $_file = '';
    if ($pear_user_config !== false) {
       $_file .= $pear_user_config;
    }
    if ($pear_system_config !== false) {
       $_file .= '/' . $pear_system_config;
    }
    if ($_file == '/') {
        $_file = 'The default config file';
    }
    $config->getMessage();
    $ui->outputData("ERROR: $_file is not a valid config file or is corrupted.");
    // We stop, we have no idea where we are :)
    exit(1);
}

// this is used in the error handler to retrieve a relative path
$_PEAR_PHPDIR = $config->get('php_dir');
$ui->setConfig($config);
PEAR::setErrorHandling(PEAR_ERROR_CALLBACK, array($ui, "displayFatalError"));
if (ini_get('safe_mode')) {
    $ui->outputData('WARNING: running in safe mode requires that all files created ' .
        'be the same uid as the current script.  PHP reports this script is uid: ' .
        @getmyuid() . ', and current user is: ' . @get_current_user());
}

$verbose = $config->get("verbose");
$cmdopts = array();

if ($raw) {
    if (!$config->isDefinedLayer('user') && !$config->isDefinedLayer('system')) {
        $found = false;
        foreach ($opts as $opt) {
            if ($opt[0] == 'd' || $opt[0] == 'D') {
                $found = true; // the user knows what they are doing, and are setting config values
            }
        }
        if (!$found) {
            // no prior runs, try to install PEAR
            if (strpos(dirname(__FILE__), 'scripts')) {
                $packagexml = dirname(dirname(__FILE__)) . DIRECTORY_SEPARATOR . 'package2.xml';
                $pearbase = dirname(dirname(__FILE__));
            } else {
                $packagexml = dirname(__FILE__) . DIRECTORY_SEPARATOR . 'package2.xml';
                $pearbase = dirname(__FILE__);
            }
            if (file_exists($packagexml)) {
                $options[1] = array(
                    'install',
                    $packagexml
                );
                $config->set('php_dir', $pearbase . DIRECTORY_SEPARATOR . 'php');
                $config->set('data_dir', $pearbase . DIRECTORY_SEPARATOR . 'data');
                $config->set('doc_dir', $pearbase . DIRECTORY_SEPARATOR . 'docs');
                $config->set('test_dir', $pearbase . DIRECTORY_SEPARATOR . 'tests');
                $config->set('ext_dir', $pearbase . DIRECTORY_SEPARATOR . 'extensions');
                $config->set('bin_dir', $pearbase);
                $config->mergeConfigFile($pearbase . 'pear.ini', false);
                $config->store();
                $config->set('auto_discover', 1);
            }
        }
    }
}
foreach ($opts as $opt) {
    $param = !empty($opt[1]) ? $opt[1] : true;
    switch ($opt[0]) {
        case 'd':
            if ($param === true) {
                die('Invalid usage of "-d" option, expected -d config_value=value, ' .
                    'received "-d"' . "\n");
            }
            $possible = explode('=', $param);
            if (count($possible) != 2) {
                die('Invalid usage of "-d" option, expected -d config_value=value, received "' .
                    $param . '"' . "\n");
            }
            list($key, $value) = explode('=', $param);
            $config->set($key, $value, 'user');
            break;
        case 'D':
            if ($param === true) {
                die('Invalid usage of "-d" option, expected -d config_value=value, ' .
                    'received "-d"' . "\n");
            }
            $possible = explode('=', $param);
            if (count($possible) != 2) {
                die('Invalid usage of "-d" option, expected -d config_value=value, received "' .
                    $param . '"' . "\n");
            }
            list($key, $value) = explode('=', $param);
            $config->set($key, $value, 'system');
            break;
        case 's':
            $store_user_config = true;
            break;
        case 'S':
            $store_system_config = true;
            break;
        case 'u':
            $config->remove($param, 'user');
            break;
        case 'v':
            $config->set('verbose', $config->get('verbose') + 1);
            break;
        case 'q':
            $config->set('verbose', $config->get('verbose') - 1);
            break;
        case 'V':
            usage(null, 'version');
        case 'c':
        case 'C':
            break;
        default:
            // all non pear params goes to the command
            $cmdopts[$opt[0]] = $param;
            break;
    }
}

if ($store_system_config) {
    $config->store('system');
}

if ($store_user_config) {
    $config->store('user');
}

$command = (isset($options[1][0])) ? $options[1][0] : null;
if (empty($command) && ($store_user_config || $store_system_config)) {
    exit;
}

if ($fetype == 'Gtk' || $fetype == 'Gtk2') {
    if (!$config->validConfiguration()) {
        PEAR::raiseError('CRITICAL ERROR: no existing valid configuration files found in files ' .
            "'$pear_user_config' or '$pear_system_config', please copy an existing configuration" .
            'file to one of these locations, or use the -c and -s options to create one');
    }
    Gtk::main();
} else do {
    if ($command == 'help') {
        usage(null, @$options[1][1]);
    }

    if (!$config->validConfiguration()) {
        PEAR::raiseError('CRITICAL ERROR: no existing valid configuration files found in files ' .
            "'$pear_user_config' or '$pear_system_config', please copy an existing configuration" .
            'file to one of these locations, or use the -c and -s options to create one');
    }

    PEAR::pushErrorHandling(PEAR_ERROR_RETURN);
    $cmd = PEAR_Command::factory($command, $config);
    PEAR::popErrorHandling();
    if (PEAR::isError($cmd)) {
        usage(null, @$options[1][0]);
    }

    $short_args = $long_args = null;
    PEAR_Command::getGetoptArgs($command, $short_args, $long_args);
    array_shift($options[1]);
    $tmp = Console_Getopt::getopt2($options[1], $short_args, $long_args);

    if (PEAR::isError($tmp)) {
        break;
    }

    list($tmpopt, $params) = $tmp;
    $opts = array();
    foreach ($tmpopt as $foo => $tmp2) {
        list($opt, $value) = $tmp2;
        if ($value === null) {
            $value = true; // options without args
        }

        if (strlen($opt) == 1) {
            $cmdoptions = $cmd->getOptions($command);
            foreach ($cmdoptions as $o => $d) {
                if (isset($d['shortopt']) && $d['shortopt'] == $opt) {
                    $opts[$o] = $value;
                }
            }
        } else {
            if (substr($opt, 0, 2) == '--') {
                $opts[substr($opt, 2)] = $value;
            }
        }
    }

    $ok = $cmd->run($command, $opts, $params);
    if ($ok === false) {
        PEAR::raiseError("unknown command `$command'");
    }

    if (PEAR::isError($ok)) {
        PEAR::setErrorHandling(PEAR_ERROR_CALLBACK, array($ui, "displayFatalError"));
        PEAR::raiseError($ok);
    }
} while (false);

// {{{ usage()

function usage($error = null, $helpsubject = null)
{
    global $progname, $all_commands;
    $stdout = fopen('php://stdout', 'w');
    if (PEAR::isError($error)) {
        fputs($stdout, $error->getMessage() . "\n");
    } elseif ($error !== null) {
        fputs($stdout, "$error\n");
    }

    if ($helpsubject != null) {
        $put = cmdHelp($helpsubject);
    } else {
        $put = "Commands:\n";
        $maxlen = max(array_map("strlen", $all_commands));
        $formatstr = "%-{$maxlen}s  %s\n";
        ksort($all_commands);
        foreach ($all_commands as $cmd => $class) {
            $put .= sprintf($formatstr, $cmd, PEAR_Command::getDescription($cmd));
        }
        $put .=
            "Usage: $progname [options] command [command-options] <parameters>\n".
            "Type \"$progname help options\" to list all options.\n".
            "Type \"$progname help shortcuts\" to list all command shortcuts.\n".
            "Type \"$progname help <command>\" to get the help for the specified command.";
    }
    fputs($stdout, "$put\n");
    fclose($stdout);

    if ($error === null) {
        exit(0);
    }
    exit(1);
}

function cmdHelp($command)
{
    global $progname, $all_commands, $config;
    if ($command == "options") {
        return
        "Options:\n".
        "     -v         increase verbosity level (default 1)\n".
        "     -q         be quiet, decrease verbosity level\n".
        "     -c file    find user configuration in `file'\n".
        "     -C file    find system configuration in `file'\n".
        "     -d foo=bar set user config variable `foo' to `bar'\n".
        "     -D foo=bar set system config variable `foo' to `bar'\n".
        "     -G         start in graphical (Gtk) mode\n".
        "     -s         store user configuration\n".
        "     -S         store system configuration\n".
        "     -u foo     unset `foo' in the user configuration\n".
        "     -h, -?     display help/usage (this message)\n".
        "     -V         version information\n";
    } elseif ($command == "shortcuts") {
        $sc = PEAR_Command::getShortcuts();
        $ret = "Shortcuts:\n";
        foreach ($sc as $s => $c) {
            $ret .= sprintf("     %-8s %s\n", $s, $c);
        }
        return $ret;

    } elseif ($command == "version") {
        return "PEAR Version: ".$GLOBALS['pear_package_version'].
               "\nPHP Version: ".phpversion().
               "\nZend Engine Version: ".zend_version().
               "\nRunning on: ".php_uname();

    } elseif ($help = PEAR_Command::getHelp($command)) {
        if (is_string($help)) {
            return "$progname $command [options] $help\n";
        }

        if ($help[1] === null) {
            return "$progname $command $help[0]";
        }

        return "$progname $command [options] $help[0]\n$help[1]";
    }

    return "Command '$command' is not valid, try '$progname help'";
}

// }}}

function error_handler($errno, $errmsg, $file, $line, $vars) {
    if ((defined('E_STRICT') && $errno & E_STRICT) || (defined('E_DEPRECATED') &&
          $errno & E_DEPRECATED) || !error_reporting()) {
        if (defined('E_STRICT') && $errno & E_STRICT) {
            return; // E_STRICT
        }
        if (defined('E_DEPRECATED') && $errno & E_DEPRECATED) {
            return; // E_DEPRECATED
        }
        if ($GLOBALS['config']->get('verbose') < 4) {
            return false; // @silenced error, show all if debug is high enough
        }
    }
    $errortype = array (
        E_ERROR   =>  "Error",
        E_WARNING   =>  "Warning",
        E_PARSE   =>  "Parsing Error",
        E_NOTICE   =>  "Notice",
        E_CORE_ERROR  =>  "Core Error",
        E_CORE_WARNING  =>  "Core Warning",
        E_COMPILE_ERROR  =>  "Compile Error",
        E_COMPILE_WARNING =>  "Compile Warning",
        E_USER_ERROR =>  "User Error",
        E_USER_WARNING =>  "User Warning",
        E_USER_NOTICE =>  "User Notice"
    );
    $prefix = $errortype[$errno];
    global $_PEAR_PHPDIR;
    if (stristr($file, $_PEAR_PHPDIR)) {
        $file = substr($file, strlen($_PEAR_PHPDIR) + 1);
    } else {
        $file = basename($file);
    }
    print "\n$prefix: $errmsg in $file on line $line\n";
    return false;
}


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * indent-tabs-mode: nil
 * mode: php
 * End:
 */
// vim600:syn=php